pipeline.py

```diff
@@ -0,0 +1,135 @@
+start_time = time.time()
+old_stdout = sys.stdout
+sys.stdout = open(os.path.join(folder_out, 'utr_annotation_temp.bed'), 'w')
+extract_three_prime_utr_information(gtf)
+sys.stdout = old_stdout
+print ('done [', round(time.time() - start_time, 2), 'seconds ]')
+
+### 1.5 Remove gene_id from utr_annotation (to be implemented in a function)
+print ("removing gene_id from 3'UTR annotation ...", end=" ", flush=True)
+start_time = time.time()
+with open(os.path.join(folder_out, 'utr_annotation_temp.bed'), 'r') as fin, open(os.path.join(folder_out, 'utr_annotation.bed'), 'w') as fout:
+    for columns in (row.strip().split('\t') for row in fin):
+        columns[3] = columns[3][16:]
+        fout.write('\t'.join(str(field) for field in columns) + '\n')
+os.remove(os.path.join(folder_out, 'utr_annotation_temp.bed'))
+print ('done [', round(time.time() - start_time, 2), 'seconds ]')
+
+### 2. Extract polyA intervals from genome
+print ('extracting polyA intervals from genome ...', end=" ", flush=True)
+if os.path.isfile(os.path.join(folder_out, 'pAi.bed')):
+    print ('skipping [ file already exists ]')
+else:
+    start_time = time.time()
+    extract_pAi_from_genome(genome, window=10, occurences=7, consecutive=6)
+    os.remove('pAi_temp.bed')
+    os.rename('pAi.bed', os.path.join(folder_out, 'pAi.bed'))
+    print ('done [', round(time.time() - start_time, 2), 'seconds ]')
+
+### 3. Add gene information to polyA intervals
+print ('adding gene annotation to pAi intervals ...', end=" ", flush=True)
+start_time = time.time()
+annotate_pAi_with_gene(os.path.join(folder_out, 'pAi.bed'), 
+                       os.path.join(folder_out, 'utr_annotation.bed'))
+os.rename('pAi_gene.bed', os.path.join(folder_out, 'pAi_gene.bed'))
+print ('done [', round(time.time() - start_time, 2), 'seconds ]')
+
+### 4. Merge polyA intervals with 3'UTRs into a dictionary
+print ("merging polyA intervals with 3'UTR ...", end=" ", flush=True)
+start_time = time.time()
+pAi_full = merge_pAi_and_utr_intervals(os.path.join(folder_out, 'utr_annotation.bed'),
+                                       os.path.join(folder_out, 'pAi_gene.bed'))
+print ('done [', round(time.time() - start_time, 2), 'seconds ]')
+
+### 5. Read bioanalyzer information
+print ('reading bioanalyzer profile ...', end=" ", flush=True)
+start_time = time.time()
+bio_size = np.array([])
+bio_intensity = np.array([])
+with open(os.path.join(folder_in, 'ds_012_50fix_bioanalyzer.txt'), 'r') as f:
+    for line in f:
+        bio_size = np.append(bio_size, int(line.split()[0]))
+        bio_intensity = np.append(bio_intensity, float(line.split()[1]))
+f_size, f_prob = discretize_bioanalyzer_profile(bio_size, bio_intensity, 5)
+print ('done [', round(time.time() - start_time, 2), 'seconds ]')
+
+### 6. Read bamfile
+print ('reading bamfile into memory ...', end=" ", flush=True)
+start_time = time.time()
+bamfile = defaultdict(list)
+with gzip.open(os.path.join(folder_in, 'ds_012_50fix_bamfile.txt.gz'), 'rt') as f:
```

Comment:
Then I suggest calling it `*.sam.gz` instead of `*.bam.txt.gz` & replace *bamfile* bu *samfile* in the code for clarity. But of course not now for testing. ;)